content/things/streams-for-reactive-programming.md

```diff
@@ -0,0 +1,411 @@
+
+```html
+<script src="https://cdn.jsdelivr.net/npm/observables-with-streams@latest/dist/really-big-bundle.js"></script>
+<script>
+  const myButton = document.querySelector("...");
+  ows.fromEvent(myButton, "click").pipeThrough(ows.map(/* ... */));
+  // ...
+</script>
+```
+
+The overall pattern I ended up with in DoF Tool is the following:
+
+```js
+// Rollup is smart enough to tree-shake despite
+// the * import.
+import * as ows from "observables-with-streams";
+
+ows
+  .combineLatest(
+    // Aperture slider
+    ows
+      .fromAsyncFunction(async () => {
+        const slider = memoizedQuerySelector("#aperture scroll-slider");
+        slider.value = (await getLastValueFromIdb("aperture")) || 2.8;
+        return ows.fromInput(slider);
+      })
+      .pipeThrough(ows.concatAll()),
+    // Focal length slider
+    ows
+      .fromAsyncFunction(async () => {
+        const slider = memoizedQuerySelector("#focal scroll-slider");
+        slider.value = (await getLastValueFromIdb("focal")) || 50;
+        return ows.fromInput(slider);
+      })
+      .pipeThrough(ows.concatAll())
+    // ... other inputs
+  )
+  .pipeThrough(ows.map(calculateFocalPlanes))
+  .pipeThrough(ows.map(calculateFieldOfView))
+  .pipeThrough(ows.forEach(adjustDOM))
+  // Only store changes in idb every second
+  .pipeThrough(ows.debounce(1000))
+  .pipeThrough(ows.forEach(storeValuesInIDB))
+  .pipeTo(ows.discard());
+```
+
+I create a couple of observables asynchronously so I can load the last values the user had from IndexedDB. Once they are initialized, I combine all my input data with `combineLatest`, `map` it through the function that do all the required math and finally use `forEach` to plug the resulting numbers back into the DOM. I also store the settings in IndexedDB, only every second or so.
+
+I have to say: I enjoyed writing an app this way. And ows held up so well that I think the experiment is a success. I didn’t encounter any problems related to the differences outlined above. Now, I know that this is not a very complex app so this verdict might not hold up at scale. YMMV. I’d love to hear from some more experienced reactive programmers what they think.
+
+### Performance
+
+Maybe somewhat surprisingly, I did not do any benchmarks. While I think there’s a good chance that RxJS will perform better in a benchmarking scenario, I have grown skeptical of the relevance of those kinds of benchmarks. The app performs well on a wide spectrum of devices (including the Nokia 2!), and that’s all that matters to me in the end. I was also more interested in comparing capabilities and developer experience first before jumping straight into the performance rabbit hole.
```

Comment:
I had expected you to mention the potential diffocult-to-measure upside of ows being inherently async and thus allowing browser scheduling to control flow rate.

Reply:
It’s microtasks, not tasks, so they still kinda block rendering. I was worried about overselling it.